assert/assertions.go

```diff
@@ -1912,33 +1920,43 @@ func EventuallyWithT(t TestingT, condition func(collect *CollectT), waitFor time
 		h.Helper()
 	}
 
-	collect := new(CollectT)
 	ch := make(chan bool, 1)
+	collect := new(CollectT)
 
-	timer := time.NewTimer(waitFor)
-	defer timer.Stop()
+	ctx, cancel := context.WithTimeout(context.Background(), waitFor)
+	defer cancel()
 
-	ticker := time.NewTicker(tick)
-	defer ticker.Stop()
+	wg := sync.WaitGroup{}
+	wg.Add(1)
 
-	for tick := ticker.C; ; {
-		select {
-		case <-timer.C:
-			collect.Copy(t)
-			return Fail(t, "Condition never satisfied", msgAndArgs...)
-		case <-tick:
-			tick = nil
+	go func() {
+		ticker := time.NewTicker(tick)
+		defer ticker.Stop()
+		defer wg.Done()
+		for {
 			collect.Reset()
-			go func() {
-				condition(collect)
-				ch <- len(collect.errors) == 0
-			}()
-		case v := <-ch:
-			if v {
-				return true
+
+			condition(collect)
+			if len(collect.errors) == 0 {
+				ch <- true
+				return
+			}
+
+			select {
+			case <-ticker.C:
+			case <-ctx.Done():
+				return
 			}
-			tick = ticker.C
 		}
+	}()
+
+	select {
+	case <-ch:
+		return true
+	case <-ctx.Done():
+		wg.Wait()
```

Comment:
This highlights a race condition that currently exists in `EventuallyWithT`: if the timer expires, we copy the CollectT's errors to the TestingT. This could happen before the goroutine executing the condition finishes, and we would not get the last tick's errors.

I think we should address this bug (#1437) prior to this PR moving forward.